File: packages/aap-android/src/main/kotlin/dev/aap/sdk/crypto/Envelope.kt
package dev.aap.sdk.crypto

import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer
import java.security.SecureRandom
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

private val rng = SecureRandom()

// ── Key generation ─────────────────────────────────────────────────────────────

/**
 * Generate an Ed25519 key pair.
 * Returns Pair(privateKeyHex, publicKeyHex).
 */
fun generateEd25519KeyPair(): Pair<String, String> {
    val privKey = Ed25519PrivateKeyParameters(rng)
    val pubKey  = privKey.generatePublicKey()
    return privKey.encoded.toHex() to pubKey.encoded.toHex()
}

/** Sign [data] with an Ed25519 private key. Returns hex-encoded signature. */
fun ed25519Sign(data: ByteArray, privateKeyHex: String): String {
    val privBytes = privateKeyHex.fromHex()
    val privKey   = Ed25519PrivateKeyParameters(privBytes, 0)
    val signer    = Ed25519Signer()
    signer.init(true, privKey)
    signer.update(data, 0, data.size)
    return signer.generateSignature().toHex()
}

/** Verify an Ed25519 signature. Returns true if valid. */
fun ed25519Verify(data: ByteArray, signatureHex: String, publicKeyHex: String): Boolean {
    return try {
        val pubBytes = publicKeyHex.fromHex()
        val sigBytes = signatureHex.fromHex()
        val pubKey   = Ed25519PublicKeyParameters(pubBytes, 0)
        val verifier = Ed25519Signer()
        verifier.init(false, pubKey)
        verifier.update(data, 0, data.size)
        verifier.verifySignature(sigBytes)
    } catch (_: Exception) { false }
}

// ── Nonce ──────────────────────────────────────────────────────────────────────

fun generateNonce(): String {
    val bytes = ByteArray(32)
    rng.nextBytes(bytes)
    return bytes.toHex()
}

// ── Envelope types ─────────────────────────────────────────────────────────────

@Serializable
data class InnerEnvelope(
    val message_id:  String,
    val action_type: String,
    val timestamp:   Long,
    val payload:     Map<String, JsonElement> = emptyMap()
)

@Serializable
data class MiddleEnvelope(
    val version:       String,
    val action_type:   String,
    val from_did:      String,
    val inner_payload: String,   // JSON string of InnerEnvelope
    val signature:     String,
    val nonce:         String
)

@Serializable
data class OuterEnvelope(
    val version:        String,
    val message_id:     String,
    val from_did:       String,
    val to_did:         String,
    val middle_payload: String,  // JSON string of MiddleEnvelope
    val signature:      String,
    val timestamp:      Long,
    val nonce:          String
)

private val json = Json { encodeDefaults = true }

/**
 * Build a three-layer AAP envelope matching the JS @aap/crypto format.
 * The JSON key ordering uses kotlinx.serialization field order which
 * mirrors the JS SDK's key ordering — required for cross-platform
 * signature verification.
 */
fun buildEnvelope(
    actionType:    String,
    payload:       Map<String, JsonElement> = emptyMap(),
    fromDID:       String,
    toDID:         String,
    privateKeyHex: String
): OuterEnvelope {
    val msgId     = java.util.UUID.randomUUID().toString()
    val timestamp = System.currentTimeMillis()

    // Inner
    val inner     = InnerEnvelope(msgId, actionType, timestamp, payload)
    val innerJSON = json.encodeToString(inner)

    // Middle — sign inner JSON
    val middleSig = ed25519Sign(innerJSON.toByteArray(), privateKeyHex)
    val middle    = MiddleEnvelope("aap/1.0", actionType, fromDID, innerJSON, middleSig, generateNonce())
    val middleJSON = json.encodeToString(middle)

    // Outer — sign middle JSON
    val outerSig = ed25519Sign(middleJSON.toByteArray(), privateKeyHex)
    return OuterEnvelope("aap/1.0", msgId, fromDID, toDID, middleJSON, outerSig, timestamp, generateNonce())
}

// ── Hex helpers ────────────────────────────────────────────────────────────────

fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }

fun String.fromHex(): ByteArray {
    require(length % 2 == 0) { "Hex string must have even length" }
    return ByteArray(length / 2) { i ->
        substring(i * 2, i * 2 + 2).toInt(16).toByte()
    }
}

File: packages/aap-swift/Sources/AAPKit/Crypto/Envelope.swift
import Foundation
import Crypto

// MARK: - Key generation

/// Generate an Ed25519 key pair. Returns (privateKeyHex, publicKeyHex).
public func generateEd25519KeyPair() -> (privateKeyHex: String, publicKeyHex: String) {
    let privateKey = Curve25519.Signing.PrivateKey()
    let pubHex  = privateKey.publicKey.rawRepresentation.hexString
    let privHex = privateKey.rawRepresentation.hexString
    return (privHex, pubHex)
}

/// Sign data with an Ed25519 private key. Returns hex-encoded signature.
public func ed25519Sign(_ data: Data, privateKeyHex: String) throws -> String {
    guard let privBytes = Data(hexString: privateKeyHex) else {
        throw AAPError.invalidPublicKey
    }
    let key = try Curve25519.Signing.PrivateKey(rawRepresentation: privBytes)
    let sig = try key.signature(for: data)
    return sig.hexString
}

/// Verify an Ed25519 signature. Returns true if valid.
public func ed25519Verify(_ data: Data, signatureHex: String, publicKeyHex: String) -> Bool {
    guard
        let pubBytes = Data(hexString: publicKeyHex),
        let sigBytes = Data(hexString: signatureHex),
        let pubKey   = try? Curve25519.Signing.PublicKey(rawRepresentation: pubBytes)
    else { return false }
    return pubKey.isValidSignature(sigBytes, for: data)
}

// MARK: - Nonce

public func generateNonce() -> String {
    var bytes = [UInt8](repeating: 0, count: 32)
    _ = SecRandomCopyBytes(kSecRandomDefault, 32, &bytes)
    return Data(bytes).hexString
}

// MARK: - Envelope builders
// These must produce JSON with the same key ordering as the JS @aap/crypto package
// so that cross-platform sessions share the same envelope format.

public struct InnerEnvelope: Codable {
    public let message_id:  String
    public let action_type: String
    public let timestamp:   Int64
    public let payload:     [String: AAPValue]

    public init(actionType: String, payload: [String: AAPValue]) {
        self.message_id  = UUID().uuidString.lowercased()
        self.action_type = actionType
        self.timestamp   = Int64(Date().timeIntervalSince1970 * 1000)
        self.payload     = payload
    }
}

public struct MiddleEnvelope: Codable {
    public let version:       String
    public let action_type:   String
    public let from_did:      String
    public let inner_payload: String   // JSON string of InnerEnvelope
    public let signature:     String
    public let nonce:         String
}

public struct OuterEnvelope: Codable {
    public let version:        String
    public let message_id:     String
    public let from_did:       String
    public let to_did:         String
    public let middle_payload: String  // JSON string of MiddleEnvelope
    public let signature:      String
    public let timestamp:      Int64
    public let nonce:          String
}

/// Build a complete three-layer AAP envelope ready to send.
public func buildEnvelope(
    actionType:        String,
    payload:           [String: AAPValue],
    fromDID:           String,
    toDID:             String,
    privateKeyHex:     String
) throws -> OuterEnvelope {
    let encoder = JSONEncoder()
    encoder.outputFormatting = .sortedKeys  // matches JS JSON.stringify key ordering

    // Inner
    let inner     = InnerEnvelope(actionType: actionType, payload: payload)
    let innerJSON = String(data: try encoder.encode(inner), encoding: .utf8)!

    // Middle — sign the inner JSON
    let innerData  = innerJSON.data(using: .utf8)!
    let middleSig  = try ed25519Sign(innerData, privateKeyHex: privateKeyHex)
    let middle     = MiddleEnvelope(
        version:       "aap/1.0",
        action_type:   actionType,
        from_did:      fromDID,
        inner_payload: innerJSON,
        signature:     middleSig,
        nonce:         generateNonce()
    )
    let middleJSON = String(data: try encoder.encode(middle), encoding: .utf8)!

    // Outer — sign the middle JSON
    let middleData = middleJSON.data(using: .utf8)!
    let outerSig   = try ed25519Sign(middleData, privateKeyHex: privateKeyHex)
    let outer      = OuterEnvelope(
        version:        "aap/1.0",
        message_id:     inner.message_id,
        from_did:       fromDID,
        to_did:         toDID,
        middle_payload: middleJSON,
        signature:      outerSig,
        timestamp:      inner.timestamp,
        nonce:          generateNonce()
    )
    return outer
}

// MARK: - Data hex helpers

extension Data {
    var hexString: String {
        map { String(format: "%02x", $0) }.joined()
    }

    init?(hexString: String) {
        let hex = hexString
        guard hex.count % 2 == 0 else { return nil }
        var data = Data(capacity: hex.count / 2)
        var idx  = hex.startIndex
        while idx < hex.endIndex {
            let next = hex.index(idx, offsetBy: 2)
            guard let byte = UInt8(hex[idx..<next], radix: 16) else { return nil }
            data.append(byte)
            idx = next
        }
        self = data
    }
}
